Harmony/Tools/AccessTools.cs

```diff
@@ -784,161 +782,520 @@ public static object[] ActualParameters(MethodBase method, object[] inputs)
+		/// Specifically, <typeparamref name="F"/> must be <see cref="Type.IsAssignableFrom(Type)">assignable from</see> OR to the field's declaring type.
+		/// Technically, this allows <c>Nullable</c>, although <c>Nullable</c> is only relevant for structs, and since only static fields of structs
+		/// are allowed for this delegate, and the instance passed to such a delegate is ignored, this hardly matters.
+		/// </para>
+		/// <para>
+		/// Similarly, <typeparamref name="F"/> is not required to be the field's field type, unless that type is a value type.
+		/// It can be a parent class (including <c>object</c>) or implemented interface of the field's field type. It cannot be a derived class.
+		/// This variance is not allowed for value types, since that would require boxing/unboxing, which is not allowed for ref values.
+		/// Specifically, for reference types, <typeparamref name="F"/> must be <see cref="Type.IsAssignableFrom(Type)">assignable from</see>
+		/// the field's field type; and for value types, <typeparamref name="F"/> must be exactly the field's field type.
+		/// </para>
+		/// <para>
+		/// This delegate supports static fields, even those defined in structs, for legacy reasons.
+		/// For such static fields, <typeparamref name="T"/> is effectively ignored.
+		/// This is also the reason that this delegate lacks a generic class constraint (it was added to certain <c>FieldRefAccess</c> methods,
+		/// but such a constraint cannot be added to this delegate without breaking binary compatibility).
+		/// Consider using <see cref="FieldRef{F}"/> (and <c>StaticFieldRefAccess</c> methods that return it) instead for static fields.
+		/// </para>
+		/// </remarks>
 		///
-		public delegate ref F FieldRef<T, F>(T obj = default);
+		public delegate ref F FieldRef<T, F>(T instance = default);
 
-		/// <summary>Creates an instance field reference</summary>
-		/// <typeparam name="T">The class the field is defined in</typeparam>
-		/// <typeparam name="F">The type of the field</typeparam>
+		/// <summary>Creates a field reference delegate for an instance field of a class</summary>
+		/// <typeparam name="T">The class that defines the instance field, or derived class of this type</typeparam>
+		/// <typeparam name="F">
+		/// The type of the field; or if the field's type is a reference type (a class or interface, NOT a struct or other value type),
+		/// a type that <see cref="Type.IsAssignableFrom(Type)">is assignable from</see> the field's type
+		/// </typeparam>
 		/// <param name="fieldName">The name of the field</param>
-		/// <returns>A read and writable field reference delegate</returns>
+		/// <returns>A readable/assignable <see cref="FieldRef{T,F}"/> delegate</returns>
 		///
-		public static FieldRef<T, F> FieldRefAccess<T, F>(string fieldName)
+		public static FieldRef<T, F> FieldRefAccess<T, F>(string fieldName) where T : class
```

Comment:
It satisfies my "binary compatibility" definition above: 

Mono ignores effectively ignores it at runtime, while .NET does check the constraint at runtime. Fortunately, all the times that it would throw, it's already "broken", satisfying my "binary compatibility definition" in the PR description:

any pre-change usage that results in an exception or crash (or undefined behavior) now results in either no exception, the same exception, or a different exception. In other words, if it was "broken" before, it either now works or throws an exception (and that thrown exception is now consistent across use cases).

The linked test logs show that this is the case across all target frameworks and .NET/mono.